Make Generic MapTo call that will generate mapping code

TylerCarlson1 opened the pull request:
This is an early work in progress but the idea is to show how using code the user writes can generate the mapping code for them automatically without the need for attributes in the most simplest of cases.  This is by no means finished and there's a lot of bugs that need to be ironed out, but the basic gist is there.

From this you only generate code based on what you actually use and not just configure.  The idea is going further that you might be able to further remove configuration settings, or at least eliminate un-used code for things that aren't being generated.
- Ex:  If you just use Projection method calls and never Map then it only generates Expression code and never map to code, since it's never being used.  But once the user writes a call to Map something, the code will then be generated for them.

### Changes
Make code using existing `MapperInformation` system and modifying to not work exclusively off of attribute mappings.
- `MapTo<,>`call will generate base code that will throw exception
- When user makes a call to it, then will auto generate `MapperInformation` for source and destination types
- The new mapping function will then add a switch statement for mapping said types and check source type and map to destination if match.

### Todos
- [x] Make switch statement only use the Types that were called from source code
  - Currently using All `MappingInformation` made, except ones with `AfterMap` because code doesn't compile.
  - No need to have all `MappingInformation`s in the switch statement if user's code won't ever reach it.
 - [x] Handle one source to multiple destinations being called in the `MapTo<,>`
  - Checks source type but not destination type for what should be returned
  - might have to move out of a switch statement depending on how it works, but uses switch right now because it's more optimized then a bunch of if statements.

- [x] Handle Configurations where both Attributes and `MapTo<,>` extension calls are used for the same `MapperInformation`
  - Currently only works for one or the other, and will double add if both are there
    - If you have Map attributes in the properties but not the Type, the system should be able to pick them up if index match
  - Need to detect if mapping already exists with Attributes and ignore, else add at the end and make sure index is +1 max mapping used from source if it doesn't already exist
  - Also need to update SyntaxReference for `MapTo<,>` validation errors calls since the source of the error isn't an attribute but a method call for the Diagnostics section
    - If use both Attributes and method calls are set will default to Attribute and errors will be shown in that location only
- [x] Figure out how to handle `MappingInformation` when using `MapTo<,>` extension method with additional constructor arguments
  - Currently blocked by simple filter for AfterMaps so can get simple scenario to work.
  - How best to notify the user, because right now the auto generated code fails with argument counts not matching exception which will confuse the user
    - I think for these cases it's probably best to use default MapToDest method calls since they will have all the parameters guaranteed.
    - Maybe tell the user a `MapTo<,>` call is using additional parameters and can't be called from there, and don't add code.
- [x] Support for Projection by method call
  - Do same thing but have method like `ProjectTo<,>` that will look for Types and generate the Expression switching between the Types
  - Maybe do the same with Enumerable mappings

ThomasBleijendaal commented on src/GeneratedMapper/MapperGenerator.cs:
```diff
@@ -56,6 +59,32 @@ public void Execute(GeneratorExecutionContext context)
+    {{
+        public static TDestination MapTo<TSource, TDestination>(this TSource source)
+        {{
+            switch (source)
```
I don't really like this switch here. If we have like 100 mappings, this switch is going to be huge. And this can lead to runtime issues when you put in a type that isn't supported, you'll get a `NotSupportedException` and no compile error. Could you generate like `MapTo<TSource, TDestination>` methods that have a `where` configure so you pin what types you can pick and the developer sees during development / build that some mapping is having a problem.

TylerCarlson1 replied:
I tried that but when I did Resharper gave ambiguous method call couldn't determine which one it was.  Maybe once the new one is created you then have to disable the main one somehow from being used for the code generated, which might be possible by I don't know how you would dynamically filter out generic conditions, so this isn't the case.
- If you know how to make it not ambiguous I'm all ears, but for now the only way I got it to work is making this class part of the code generation process.
- Need to start with something people can call, and when they call add on top of it so their case works.

The idea is that if you call `MapTo<TSource,TDestination>` at all it will make a case for said types no matter what, so you'll never hit the runtime not supported exception.  The issue comes into play adding Maps that have AfterMap or Extensions with constructor parameters and how we inform the user.

1. Put the code into the switch statement and let the compilation fail with Argument count miss-match happen in the MappingExpression.g.cs
2. Identify the miss match in validation and then add Error to diagnostics explaining the resolvers/extensions are blocking them from using code in `MapTo<TSource,TDestination>` function, and have the source be the call in question vs the auto generated code.
  - Also question of putting the code in there as well, so you will get both exceptions, but the first will be the custom one we make which they should solve first.  That's usually how compilers work.  Fix first thing and fixes a bunch of other smaller issues that were inferred incorrectly from the original problem.

ThomasBleijendaal replied:
Yeah after giving it more thought I realized that this is a C# syntax limitation. You can't make overloads where only the `where`s are different. So this is something we have to accept.

For the issue with the parameters, that is going to be quite a challenge. You have one single `MapTo<,>` name you want to use for potentially all the mapping methods, which could have a large variety of parameter combinations. Solving that is perhaps a bit too much for the initial implementation. Perhaps we should view this `MapTo<,>` syntax for now an ad-hoc variant, where there is no support for parameters yet.

ThomasBleijendaal commented on src/GeneratedMapper/SyntaxReceivers/MapAttributeReceiver.cs:
```diff
@@ -9,6 +10,9 @@ namespace GeneratedMapper.SyntaxReceivers
         public List<TypeDeclarationSyntax> Candidates { get; } = new List<TypeDeclarationSyntax>();
+        private List<TypeDeclarationSyntax> _cachedSyntaxes = new List<TypeDeclarationSyntax>();
+        public List<Tuple<TypeDeclarationSyntax, TypeDeclarationSyntax>> ExtensionCandidates { get; } = new List<Tuple<TypeDeclarationSyntax, TypeDeclarationSyntax>>();
+        private List<Tuple<TypeSyntax, TypeSyntax>> _extensionCandidates = new List<Tuple<TypeSyntax, TypeSyntax>>();
```
I think its better to use value types instead of `Tuple<>`, so `List<(TypeSyntax source, TypeSyntax destination)>`

TylerCarlson1 replied:
Agreed.  This was simple quick and dirty way of getting it to work.
I might have to throw in SyntaxReference object in the `ExtensionCandiates` so I can know where the source is, since it's not always an attribute that is the source of the mapping code generation.

TylerCarlson1 replied:
Did make ExtensionMapCandidate to fix this issue already.

ThomasBleijendaal commented:
Another thing to consider is how this `MapTo<,>` is going to work when it is used in generic environments. Let's say we have a generic repository which maps a model to and from the database entity, using this generic `MapTo<,>`. Will we be able to generate the correct mappers based on this use? It would be valuable, because the current `MapToConcreteTarget` method name prohibits use in generics.

TylerCarlson1 commented:
Fixed a few things.
- If you have after map and use `MapTo<,>` it will give argument doesn't have override of 1 parameter that will show the error in MappingExtensions.g.cs.
  - I don't know if you can come up with a better idea right now other than leave it as is and see how much of a problem it actually is.
- With the switch statement for destination it doesn't work for types inside types yet.  FullName has a + and doesn't give that by default.
- Only applies MapToDestination code and validation to first MapConfiguration if multiple exist.
  - Index priority with Extensions being last as they can't have any configuration.
  - Before would probably throw error since generating the same file twice.

Still need to figure out how to get the `SyntaxReference` for GenericMethod call vs Attribute.
Once I get that I can then reference generic methods that call it and reference syntax back to that location.
Then last thing is Project To and then onto refactoring and cleaning up the code gen to use IndexWriter and such.

ThomasBleijendaal commented on src/GeneratedMapper/Information/MappingInformation.cs:
```diff
@@ -13,17 +13,59 @@ internal sealed class MappingInformation
+                    return true;
+                }
+
+                if (ReferenceEquals(x, null))
```
This looks a bit silly. Isn't a `if (x == null || y == null) { return false; }` more easy?

TylerCarlson1 replied:
Resharper made this code and got the job done.  I'll worry about looks and placement in the refactor.

TylerCarlson1 replied:
Fixed

ThomasBleijendaal commented on src/GeneratedMapper/MapperGenerator.cs:
```diff
@@ -39,23 +42,55 @@ public void Execute(GeneratorExecutionContext context)
-                ValidateMappings(context, foundMappings);
+                var distinctMappings = foundMappings
+                    .GroupBy(x => x, MappingInformation.SourceTypeDestinationTypeComparer)
+                    .Select(x => x.OrderByDescending(m => m.AttributeIndex).First()).ToArray();
```
Why have you chosen for picking the highest index? Don't you feel like that `Index = 1` would be the best match?

TylerCarlson1 replied:
Extensions index to `Int32.MinValue`, so I didn't want that to be the first, but yea it could be max and then could `OrderBy` making first be 1.  I just thought negative index's for extensions would differentiate them from attribute index's and made code work off that assumption.

I'll probably change it so index 1 is first and make index max for `MapTo<,>` calls.

TylerCarlson1 replied:
Fixed already

ThomasBleijendaal commented on src/GeneratedMapper/SyntaxReceivers/ExtensionMapCandidate.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace GeneratedMapper.SyntaxReceivers
+{
+    public class ExtensionMapCandidate
```
Probably make this also `internal`

TylerCarlson1 replied:
Already Changed to Internal and sealed

TylerCarlson1 commented:
I've added ProjectTo and it will generate the Expression methods if needed.  Also when just calling 'MapTo<,>' it won't generate the enumerable anymore since it's not being used.
For Extension calls I got Syntax location to show up properly in the diagnostics.

As for the generic method calling of things like IRepository I don't think I can guarantee that, or it's super cumbersome to deal with right now, because as my understanding goes it only addresses syntax once.  That means you have to cache every method call that uses generics and then if you get generic extension call, then you have to back port the type map configurations seeing if the method was ever called and get those types.  Problem is when you use generics there's no way to know it's generic at the time since it's just syntax.  You have to figure that out from Type Arguments from the function that called it and then infer back on it.  It's probably possible but it's way more involved on top of all the other things changing.

My question is how critical is it that this feature has that kind of detection right now?
- Current system will throw compile time exceptions saying type doesn't exist, and code gen fails
- This would be put into the readme instructions saying, generics like IRepository don't work right now.

So I think I'm done functionality wise, so I have a few questions for you.
1. Have any questions on things work, or want to be refactored to be more easily readable?
   - I did re-write the Expression generation code to use indent writer so it's easier to follow and matches the rest of the patterns
   - MappingType is now bit flagged and checking flags, and compounding multiple configurations calls to selectively generate map and expression files.
2.  To what extent do you need this to be unit tested for it to be acceptable?
    - I see you have a lot of tests with existing code, so I would like your opinion on this.
    - There are many scenarios as far as this goes to a pretty big degree
      - `[MapTo(typeof())]`,`[MapFrom(typeof())]`,`MapTo<,>`,`ProjectTo<,>` in various mix-match configurations
      - Also failures and making sure the message and location goes to the correct place.
      - If have attributes and projection, should default to Attribute as error message and possibly not show projection issue
      - If projection error should use correct error code saying that attribute mappings are needed to configure the correction.

ThomasBleijendaal commented:
I don't think it is important to have the generics totally correct in the first implementation. I can see that those generics will become quite a challenge to properly support, especially when class hierarchies come in to play. (Map A-B, A-B', A'-B', etc..)

Furthermore, for these use case we can suggest to use the injectable mappers (`IMapper<,>`) instead and have them be hard-configured with attributes. It can be useful to have the `IMapper<,>` also be autogenerated, just like how `.MapTo<,>` works now, but this is all stuff for other PRs. I'll create issues just to track these ideas.

I'll review your code in the next few days, and see what I miss in the current unit test. I have a lot of unit tests in place, but lots of those tests are quite redundant. If the tests in this PR test how this pattern gets tied in to the generator its probably enough for now, no need to test all the possible options. I do want to see tests for what diagnostics are outputted, but I'll let you know.

ThomasBleijendaal commented on src/GeneratedMapper/SyntaxReceivers/MapAttributeReceiver.cs:
```diff
@@ -8,10 +9,15 @@ namespace GeneratedMapper.SyntaxReceivers
     internal sealed class MapAttributeReceiver : ISyntaxReceiver
     {
-        public List<TypeDeclarationSyntax> Candidates { get; } = new List<TypeDeclarationSyntax>();
+        public List<TypeDeclarationSyntax> Candidates { get; } = new();
```
I think we need to rename this to `AttributeCandidates`.

TylerCarlson1 replied:
Ok can do

TylerCarlson1 replied:
Didn't update this because it is originally Attribute candidates but I do append projection and extension candidates in there too with Trim, so it' all candidates.

ThomasBleijendaal reviewed (changes requested):
There are some null references here and there that need fixing. Most of the code is fine, there are a few suggestions and I think some unit tests need to be added, especially for the diagnostics.

TylerCarlson1 commented:
Added unit tests to test functionality in both in compilation and diagnostics testing.
- I added the MapExtensions.g.cs and ProjectToExtensions.g.cs files to Output tester help method.
  - If 2 files extra in output test them against default file
  - If 1 extra file test against ProjectTo default file.
- Made new test function for location text for error, to test the Extension methods point to `MapTo<,>` and `ProjectTo<,>` location
- Tested MapTo/MapFrom attributes override `MapTo<,>` and `ProjectTo<,>` no configuration setting
  - The more advanced first or highest index testing wasn't added for things like having both Attributes and sorting by Index

So I think it's done from my part.
- For null references I'm not 100% sure, so you can tell me or update the branch (You should have access to push and pull from this branch to fix things if you want)
- If I missed any testing scenarios tell me and I'll be sure to add them.

ThomasBleijendaal commented on src/GeneratedMapper.Tests/Helpers/GeneratorTestHelper.cs:
```diff
@@ -43,6 +81,20 @@ public static void TestGeneratedCode(string sourceText, params string[] expected
                 Assert.AreEqual(expectedOutputSourceTexts[i], output.ElementAtOrDefault(i) ?? "", $"Error in file index: {i}");
             }
+
+            if (output.Length == expectedOutputSourceTexts.Length + 2)
```
I understand that you do this, but this can become a bit difficult when we have another feature added that might generate another file. Could you fix this with adding some flag enum which configures which default text must be added, and give it a default of `GeneratorDefaults.MapExtensions | GeneratorDefaults.ProjectToExtensions`.

TylerCarlson1 replied:
I didn't use enum since params are being used and can't have a default and params without changing all the unit tests.
What I did instead was made a DefaultFileText list, and then get count difference between expected and actual outputs and push the files onto the end so the counts match up.

This way if more files are added to the list we just need to insert the new default into DefaultFileText and it should properly check all unit tests without any additional test code changes.

ThomasBleijendaal replied:
Oh yeah that is right. It's ok for now, but I can see some issues down the road when we have more files generated down the road and someone wants to specifically test one of those files. But we'll cross that bridge when we get there.

ThomasBleijendaal commented on src/GeneratedMapper.Tests/ProjectToExtensionMethodTests.cs:
```diff
@@ -0,0 +1,95 @@
+                    {
+                        ""Test.B"" =>
+                            a.Select(Test.Expressions.A.ToB()) is IQueryable<TDestination> b ? b : default,
+                        _ => throw new NotSupportedException(""Projection is not configured"")
```
Can you add what projection is not configured, that could help with debugging.

TylerCarlson1 replied:
Added for both Projection and Map using TSource and TDestination.  They should never be able to hit this because compilation should fail before it gets this far, at least for now, but better safe then sorry.

ThomasBleijendaal commented on src/GeneratedMapper/Builders/ProjectToExtensionsBuilder.cs:
```diff
@@ -0,0 +1,77 @@
+
+namespace GeneratedMapper.Builders
+{
+    internal sealed class ProjectToExtensionsBuilder
```
This object is pretty similar to `MapToExtensionsBuilder`. I think most shared coded can be put in a base class and only have the mapto or projectto specifics be in the derived classes.

TylerCarlson1 replied:
I put a small base with not that much code share between them.  Only really the IndentWriter creation, which is so commonly used it should probably be its own Extension method so you don't repeat conditional operation in its constructor.
Maybe a wrapper that passes indent writer but can dispose of the String writer at the same time since the indent writer doesn't dispose of string writer when its Disposed.  That way there's only one writer instead of two at the start of writing files.

As for sharing other code I didn't want to attempt it since it would break up the flow of the spacing matching code if you hid it.  Sure you could refactor the code in the foreach loops into easier to read functions to make the indentWriter.WriteLine more readable, but how much does it really help?

ThomasBleijendaal replied:
Ok. I'll have a look at after merging, see if I can perhaps improve on it.

TylerCarlson1 commented:
@ThomasBleijendaal You approved it but didn't resolve the conversations.  Is there still something I need to change in order to submit it?

ThomasBleijendaal commented:
@TylerCarlson1 There are still a few things that I left open, but if you fix those I'll merge the PR directly.

- https://github.com/ThomasBleijendaal/GeneratedMapper/pull/31/files/156f3f5ed99df66d5540db9d7b2b84c14aa6eee1#r674219167
- https://github.com/ThomasBleijendaal/GeneratedMapper/pull/31/files/156f3f5ed99df66d5540db9d7b2b84c14aa6eee1#r674217623
- https://github.com/ThomasBleijendaal/GeneratedMapper/pull/31/files#r674215162
- https://github.com/ThomasBleijendaal/GeneratedMapper/pull/31/files/22874ad366e2eb5aaed99db318f334a7e24e25e5#r671821559
